server/server.go

```diff
@@ -18,63 +18,79 @@ package server
 	"fmt"
-	"sync"
-	"time"
 
 	"metacontroller.io/controller/common"
 
 	"metacontroller.io/controller/decorator"
 	"metacontroller.io/options"
 
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
 	"metacontroller.io/apis/metacontroller/v1alpha1"
 	mcclientset "metacontroller.io/client/generated/clientset/internalclientset"
 	"metacontroller.io/controller/composite"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
 
 	_ "k8s.io/client-go/plugin/pkg/client/auth/oidc"
 )
 
-type controller interface {
-	Start()
-	Stop()
-}
+func New(configuration options.Configuration) (controllerruntime.Manager, func(), error) {
+	mgr, err := controllerruntime.NewManager(configuration.RestConfig, manager.Options{
+		// Disables serving built-in metrics.
+		// We already start a standalone metrics server in parallel to the manager.
+		MetricsBindAddress: "0",
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	err = v1alpha1.AddToScheme(mgr.GetScheme())
+	if err != nil {
+		return nil, nil, err
+	}
 
-func Start(configuration options.Configuration) (stop func(), err error) {
 	// Create informer factory for metacontroller API objects.
 	mcClient, err := mcclientset.NewForConfig(configuration.RestConfig)
 	if err != nil {
-		return nil, fmt.Errorf("can't create client for api %s: %v", v1alpha1.SchemeGroupVersion, err)
+		return nil, nil, fmt.Errorf("can't create client for api %s: %v", v1alpha1.SchemeGroupVersion, err)
 	}
 
 	controllerContext, err := common.NewControllerContext(configuration, mcClient)
-
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
+	controllerContext.Start()
 
-	controllers := []controller{
-		composite.NewMetacontroller(*controllerContext, mcClient, configuration.Workers),
-		decorator.NewMetacontroller(*controllerContext, configuration.Workers),
+	compositeReconciler := composite.NewMetacontroller(*controllerContext, mcClient, configuration.Workers)
+	compositeCtrl, err := controller.New("composite-metacontroller", mgr, controller.Options{
+		Reconciler: compositeReconciler,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	err = compositeCtrl.Watch(&source.Kind{Type: &v1alpha1.CompositeController{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return nil, nil, err
 	}
 
-	controllerContext.Start()
-
-	// Start all controllers.
-	for _, c := range controllers {
-		c.Start()
+	decoratorReconciler := decorator.NewMetacontroller(*controllerContext, configuration.Workers)
+	decoratorCtrl, err := controller.New("decorator-metacontroller", mgr, controller.Options{
+		Reconciler: decoratorReconciler,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	err = decoratorCtrl.Watch(&source.Kind{Type: &v1alpha1.DecoratorController{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return nil, nil, err
 	}
 
-	// Return a function that will stop all controllers.
```

Comment:
It is usually a good practice to decouple initiating and starting managers. So I decided to move the logic to start and terminate the manager into main.go